src/internal/pfsdb/commits.go

```diff
@@ -0,0 +1,956 @@
+	rev     bool
+	limit   int
+	offset  int
+	commits []commitRow
+	index   int
+	filter  CommitListFilter
+}
+
+// Next advances the iterator by one row. It returns a stream.EOS when there are no more entries.
+func (iter *CommitIterator) Next(ctx context.Context, dst *CommitPair) error {
+	return errors.Wrap(dbutil.WithTx(ctx, iter.db, func(ctx context.Context, tx *pachsql.Tx) error {
+		return iter.next(ctx, tx, dst)
+	}, dbutil.WithReadOnly()), "next commit")
+}
+
+func (iter *CommitIteratorTx) Next(ctx context.Context, dst *CommitPair) error {
+	return errors.Wrap(iter.next(ctx, iter.tx, dst), "next commit in transaction")
+}
+
+func (iter *commitIterator) next(ctx context.Context, tx *pachsql.Tx, dst *CommitPair) error {
+	if dst == nil {
+		return errors.Wrap(fmt.Errorf("commit is nil"), "get next commit")
+	}
+	var err error
+	if iter.index >= len(iter.commits) {
+		iter.index = 0
+		iter.offset += iter.limit
+		iter.commits, err = listCommitPage(ctx, tx, iter.limit, iter.offset, iter.filter, iter.rev)
+		if err != nil {
+			return errors.Wrap(err, "list commit page")
+		}
+		if len(iter.commits) == 0 {
+			return stream.EOS()
+		}
+	}
+	row := iter.commits[iter.index]
+	commit, err := getCommitInfoFromCommitRow(ctx, tx, &row)
+	if err != nil {
+		return errors.Wrap(err, "getting commitInfo from commit row")
+	}
+
+	*dst = CommitPair{
+		CommitInfo: commit,
+		ID:         row.RowID,
+	}
+	iter.index++
+	return nil
+}
+
+// CommitFields is used in the ListCommitFilter and defines specific field names for type safety.
+// This should hopefully prevent a library user from misconfiguring the filter.
+type CommitFields string
+
+var (
+	CommitSetIDs   = CommitFields("commit_set_id")
+	CommitOrigins  = CommitFields("origin")
+	CommitRepos    = CommitFields("repo_id")
+	CommitBranches = CommitFields("branch_id")
+	CommitProjects = CommitFields("project_id")
+)
+
+// CommitListFilter is a filter for listing commits. It ANDs together separate keys, but ORs together the key values:
+// where commit.<key_1> IN (<key_1:value_1>, <key_2:value_2>, ...) AND commit.<key_2> IN (<key_2:value_1>,<key_2:value_2>,...)
+type CommitListFilter map[CommitFields][]string
+
+// ListCommit returns a CommitIterator that exposes a Next() function for retrieving *pfs.CommitInfo references.
+func ListCommit(ctx context.Context, db *pachsql.DB, filter CommitListFilter, rev bool) (*CommitIterator, error) {
+	var iter *commitIterator
+	var err error
+	if err := dbutil.WithTx(ctx, db, func(cbCtx context.Context, tx *pachsql.Tx) error {
+		iter, err = newIterator(ctx, tx, filter, rev)
+		return err
+	}, dbutil.WithReadOnly()); err != nil {
+		return nil, errors.Wrap(err, "list commits")
+	}
+	return &CommitIterator{iter, db}, nil
+}
+
+// ListCommitTx returns a CommitIterator that exposes a Next() function for retrieving *pfs.CommitInfo references.
+func ListCommitTx(ctx context.Context, tx *pachsql.Tx, filter CommitListFilter, rev bool) (*CommitIteratorTx, error) {
```

Comment:
I like the pattern of suffixing these with `Tx`, we use `InTransaction` elsewhere, and it's just too many letters.

Reply:
This is a dangerous method to make public.  I can't think of any legitimate use cases.   All the use cases that come to mind would be better off using Select.

Reply:
I actually have a use case for this in our unit tests. We want to verify that all 18 of our test commits are migrated and I can use this there. That said, the regular ListCommit() will work fine too.

Reply:
Okay if you're using it in a test, that's fine, can we make it private?